In [1]:
import re
import pandas as pd
pd.options.display.max_rows = 999

In [2]:
!ls /Users/clare.huang/AskGene/data/external/*.txt

/Users/clare.huang/AskGene/data/external/lung_cancer_terminology.txt


In [3]:
fname = '/Users/clare.huang/AskGene/data/external/lung_cancer_terminology.txt'
with open(fname, 'r') as f:
    lines = f.readlines()

with open(fname, 'r') as f:
    lines_copy = f.read()


In [4]:
with open('rows_and_cols_terminology.txt', 'w') as f:
    for line in lines:
        title_check = re.search('[\d]{3} [\d]{4}/[\d] (.*) \n', line)
        subline_check = re.search('[\d]{4}/[\d] (.*) \n', line)
        if title_check:
            f.write(title_check.group(0))
            title_num = title_check.group(0)[:4]
            lines_copy = re.sub(title_check.group(0), '', lines_copy)
        elif subline_check:
            f.write(title_num + subline_check.group(0))
            lines_copy = re.sub(subline_check.group(0), '', lines_copy)        
        else:
            print(line)


Page 1/7

ICD-0-3 SEER SITE/HISTOLOGY VALIDATION LIST September 18, 2015 143 







LUNG & BRONCHUS C340-C343,C348-C349 



NEOPLASM 





CARCINOMA, NOS 







CARCINOMA, UNDIFF., NOS 

GIANT & SPINDLE CELL CARCINOMA 





http://seer.cancer.gov/icd-o-3/143 of 381



Hide Page 2/7

ICD-0-3 SEER SITE/HISTOLOGY VALIDATION LIST September 18, 2015 144 







LUNG & BRONCHUS C340-C343,C348-C349 



SQUAMOUS CELL CARCINOMA, NOS 













TRANSITIONAL CELL CARCINOMA, NOS 





ADENOCARCINOMA, NOS 



ADENOID CYSTIC & CRIBRIFORM CA. 

SOLID CARCINOMA, NOS 

CARCINOID TUMOR, MALIGNANT 

http://seer.cancer.gov/icd-o-3/144 of 381



Hide Page 3/7

ICD-0-3 SEER SITE/HISTOLOGY VALIDATION LIST September 18, 2015 145 





LUNG & BRONCHUS C340-C343,C348-C349 

http://seer.cancer.gov/icd-o-3/145 of 381



Hide Page 4/7

ICD-0-3 SEER SITE/HISTOLOGY VALIDATION LIST September 18, 2015 146 







LUNG & BRONCHUS C340-C343,C348-C349 















http://seer.cancer.gov/icd-o-3/146 of 381



Hide

In [5]:
with open('big_topic.txt','w') as f:
    for line in lines_copy.split('\n'):
        if len(line) == 0:
            continue
        if 'Page' in line:
            continue
        elif 'LUNG & BRONCHUS C' in line:
            continue
        elif 'SEER SITE/HISTOLOGY VALIDATION LIST' in line:
            continue
        else:
            line = re.sub(r'http://seer.cancer.gov/icd-o-3/[\d]{3} of [\d]{3}', '', line)
        print(line)
        f.write(line + '\n')


NEOPLASM 
CARCINOMA, NOS 
CARCINOMA, UNDIFF., NOS 
GIANT & SPINDLE CELL CARCINOMA 
SMALL CELL CARCINOMA, NOS NON-SMALL CELL CARCINOMA, NOS PAPILLARY CARCINOMA, NOS 
SQUAMOUS CELL CARCINOMA, NOS 
LYMPHOEPITHELIAL CARCINOMA 
TRANSITIONAL CELL CARCINOMA, NOS 
ADENOCARCINOMA, NOS 
ADENOID CYSTIC & CRIBRIFORM CA. 
SOLID CARCINOMA, NOS 
CARCINOID TUMOR, MALIGNANT 

CARCINOID TUMOR, MALIGNANT BRONCHIOLO-ALVEOLAR ADENOCA. PAPILLARY ADENOCARCINOMA, NOS CLEAR CELL ADENOCARCINOMA, NOS GRANULAR CELL CARCINOMA MUCOEPIDERMOID CARCINOMA MUCINOUS ADENOCARCINOMA SIGNET RING CELL CARCINOMA MEDULLARY CARCINOMA, NOS ACINAR CELL CARCINOMA ADENOSQUAMOUS CARCINOMA ADENOCA. WITH METAPLASIA 
ADENOCA. WITH METAPLASIA SARCOMA, NOS FIBROMATOUS NEOPLASMS FIBROUS HISTIOCYTOMA, MAL. MYOMATOUS NEOPLASMS RHABDOMYOSARCOMA, NOS EMBRYONAL RHABDOMYOSARCOMA HEPATOBLASTOMA CARCINOSARCOMA, NOS 

MESENCHYMOMA, MALIGNANT 
899 
8990/3 
Mesenchymoma, malignant 
MESOTHELIOMA, MALIGNANT BLOOD VESSEL TUMORS HEMANGIOENDOTHELIOMA KAP

In [6]:
with open('rows_and_cols_terminology.txt', 'r') as f:
    load_cols = f.readlines()

make_df = list()
for line in load_cols:
    #print((line[:3], line[4:10], line[11:].rstrip()))
    make_df.append((line[:3], line[4:10], line[11:].rstrip()))

df = pd.DataFrame(make_df, columns=['Category_code', 'Sub-category code', 'Sub-category'])

In [7]:
with open('big_topic_hand_edited.txt', 'r') as f:
    big_topics = f.readlines()
big_topics = [x.rstrip() for x in big_topics]    


### Manually remove an entry leading to column mismatch

In [8]:
print(big_topics.index('NON-SMALL CELL CARCINOMA, NOS'))
big_topics.remove('NON-SMALL CELL CARCINOMA, NOS')
big_topics.remove('MESENCHYMOMA, MALIGNANT')

5


In [9]:
unique_levels = df['Category_code'].unique().tolist()
unique_levels = unique_levels + ['997']
unique_levels = [int(x) for x in unique_levels]

print('Size of unique_level: \n', len(unique_levels))
corresponding_dict = dict(zip(sorted(unique_levels), big_topics))
corresponding_dict['889'] = 'MYOMATOUS NEOPLASMS'
corresponding_dict['899'] = 'MESENCHYMOMA, MALIGNANT'
corresponding_dict['982'] = 'LYMPHOID LEUKEMIA, NOS'

for key in corresponding_dict:
    print(key, corresponding_dict[key])

Size of unique_level: 
 54
800 NEOPLASM
801 CARCINOMA, NOS
802 CARCINOMA, UNDIFF., NOS
803 GIANT & SPINDLE CELL CARCINOMA
804 SMALL CELL CARCINOMA, NOS
805 PAPILLARY CARCINOMA, NOS
807 SQUAMOUS CELL CARCINOMA, NOS
808 LYMPHOEPITHELIAL CARCINOMA
812 TRANSITIONAL CELL CARCINOMA, NOS
814 ADENOCARCINOMA, NOS
820 ADENOID CYSTIC & CRIBRIFORM CA.
823 SOLID CARCINOMA, NOS
824 CARCINOID TUMOR, MALIGNANT
825 BRONCHIOLO-ALVEOLAR ADENOCA.
826 PAPILLARY ADENOCARCINOMA, NOS
831 CLEAR CELL ADENOCARCINOMA, NOS
832 GRANULAR CELL CARCINOMA
843 MUCOEPIDERMOID CARCINOMA
848 MUCINOUS ADENOCARCINOMA
849 SIGNET RING CELL CARCINOMA
851 MEDULLARY CARCINOMA, NOS
855 ACINAR CELL CARCINOMA
856 ADENOSQUAMOUS CARCINOMA
857 ADENOCA. WITH METAPLASIA
880 SARCOMA, NOS
881 FIBROMATOUS NEOPLASMS
883 FIBROUS HISTIOCYTOMA, MAL.
889 MYOMATOUS NEOPLASMS
890 RHABDOMYOSARCOMA, NOS
891 EMBRYONAL RHABDOMYOSARCOMA
897 HEPATOBLASTOMA
898 CARCINOSARCOMA, NOS
905 MESOTHELIOMA, MALIGNANT
912 BLOOD VESSEL TUMORS
913 HEMANGIOENDOTHELIO

In [10]:
len(big_topics)

53

# Fix any mismatch found

In [11]:
df['Category'] = df['Category_code'].map(lambda x: corresponding_dict[int(x)])

## Manually fix the entry of NON-SMALL CELL CARCINOMA, NOS (804)

In [12]:
df.loc[27,'Category'] = 'NON-SMALL CELL CARCINOMA, NOS'

In [13]:
df = df[['Category', 'Category_code', 'Sub-category code', 'Sub-category']]

In [14]:
df.to_csv('parsed_histology_validation_list.csv', index=None)

In [15]:
df

Category Category_code  \
0                                             NEOPLASM           800   
1                                             NEOPLASM           800   
2                                             NEOPLASM           800   
3                                             NEOPLASM           800   
4                                             NEOPLASM           800   
5                                             NEOPLASM           800   
6                                       CARCINOMA, NOS           801   
7                                       CARCINOMA, NOS           801   
8                                       CARCINOMA, NOS           801   
9                                       CARCINOMA, NOS           801   
10                                      CARCINOMA, NOS           801   
11                                      CARCINOMA, NOS           801   
12                                      CARCINOMA, NOS           801   
13                             CARCINOMA, UNDIFF., NOS           802   
14                             CARCINOMA, UNDIFF., NOS           802   
15                             CARCINOMA, UNDIFF., NOS           802   
16                      GIANT & SPINDLE CELL CARCINOMA           803   
17                      GIANT & SPINDLE CELL CARCINOMA           803   
18                      GIANT & SPINDLE CELL CARCINOMA           803   
19                      GIANT & SPINDLE CELL CARCINOMA           803   
20                      GIANT & SPINDLE CELL CARCINOMA           803   
21                      GIANT & SPINDLE CELL CARCINOMA           803   
22                           SMALL CELL CARCINOMA, NOS           804   
23                           SMALL CELL CARCINOMA, NOS           804   
24                           SMALL CELL CARCINOMA, NOS           804   
25                           SMALL CELL CARCINOMA, NOS           804   
26                           SMALL CELL CARCINOMA, NOS           804   
27                       NON-SMALL CELL CARCINOMA, NOS           804   
28                            PAPILLARY CARCINOMA, NOS           805   
29                            PAPILLARY CARCINOMA, NOS           805   
30                            PAPILLARY CARCINOMA, NOS           805   
31                            PAPILLARY CARCINOMA, NOS           805   
32                            PAPILLARY CARCINOMA, NOS           805   
33                        SQUAMOUS CELL CARCINOMA, NOS           807   
34                        SQUAMOUS CELL CARCINOMA, NOS           807   
35                        SQUAMOUS CELL CARCINOMA, NOS           807   
36                        SQUAMOUS CELL CARCINOMA, NOS           807   
37                        SQUAMOUS CELL CARCINOMA, NOS           807   
38                        SQUAMOUS CELL CARCINOMA, NOS           807   
39                        SQUAMOUS CELL CARCINOMA, NOS           807   
40                        SQUAMOUS CELL CARCINOMA, NOS           807   
41                        SQUAMOUS CELL CARCINOMA, NOS           807   
42                        SQUAMOUS CELL CARCINOMA, NOS           807   
43                    TRANSITIONAL CELL CARCINOMA, NOS           812   
44                    TRANSITIONAL CELL CARCINOMA, NOS           812   
45                    TRANSITIONAL CELL CARCINOMA, NOS           812   
46                    TRANSITIONAL CELL CARCINOMA, NOS           812   
47                    TRANSITIONAL CELL CARCINOMA, NOS           812   
48                    TRANSITIONAL CELL CARCINOMA, NOS           812   
49                                 ADENOCARCINOMA, NOS           814   
50                                 ADENOCARCINOMA, NOS           814   
51                                 ADENOCARCINOMA, NOS           814   
52                                 ADENOCARCINOMA, NOS           814   
53                                 ADENOCARCINOMA, NOS           814   
54                     ADENOID CYSTIC & CRIBRIFORM CA.           820   
55            